File: src/main/java/net/ccbluex/liquidbounce/launch/data/legacyui/ClickGUIModule.java
/*
 * FDPClient Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge by LiquidBounce.
 * https://github.com/UnlegitMC/FDPClient/
 */
package net.ccbluex.liquidbounce.launch.data.legacyui;

import net.ccbluex.liquidbounce.event.EventTarget;
import net.ccbluex.liquidbounce.event.PacketEvent;
import net.ccbluex.liquidbounce.features.module.Module;
import net.ccbluex.liquidbounce.features.module.ModuleCategory;
import net.ccbluex.liquidbounce.features.module.ModuleInfo;
import net.ccbluex.liquidbounce.launch.data.legacyui.clickgui.ClickGui;
import net.ccbluex.liquidbounce.launch.data.legacyui.clickgui.style.styles.*;
import net.ccbluex.liquidbounce.launch.options.LegacyUiLaunchOption;
import net.ccbluex.liquidbounce.utils.render.ColorUtils;
import net.ccbluex.liquidbounce.value.BoolValue;
import net.ccbluex.liquidbounce.value.FloatValue;
import net.ccbluex.liquidbounce.value.IntegerValue;
import net.ccbluex.liquidbounce.value.ListValue;
import net.minecraft.network.Packet;
import net.minecraft.network.play.server.S2EPacketCloseWindow;
import org.lwjgl.input.Keyboard;

import java.awt.*;

@ModuleInfo(name = "ClickGUI", category = ModuleCategory.CLIENT, keyBind = Keyboard.KEY_RSHIFT, canEnable = false)
public class ClickGUIModule extends Module {
    private final ListValue styleValue = new ListValue("Style", new String[] {"LiquidBounce", "Null", "Slowly", "Black", "White"}, "Slowly") {
        @Override
        protected void onChanged(final String oldValue, final String newValue) {
            updateStyle();
        }
    };

    public final FloatValue scaleValue = new FloatValue("Scale", 1F, 0.7F, 2F);
    public final IntegerValue maxElementsValue = new IntegerValue("MaxElements", 15, 1, 20);

    private static final BoolValue colorRainbow = new BoolValue("Rainbow", false);
    private static final IntegerValue colorRedValue = (IntegerValue) new IntegerValue("R", 0, 0, 255).displayable(() -> !colorRainbow.get());
    private static final IntegerValue colorGreenValue = (IntegerValue) new IntegerValue("G", 160, 0, 255).displayable(() -> !colorRainbow.get());
    private static final IntegerValue colorBlueValue = (IntegerValue) new IntegerValue("B", 255, 0, 255).displayable(() -> !colorRainbow.get());

    public static Color generateColor() {
        return colorRainbow.get() ? ColorUtils.INSTANCE.rainbow() : new Color(colorRedValue.get(), colorGreenValue.get(), colorBlueValue.get());
    }

    @Override
    public void onEnable() {
        updateStyle();

        mc.displayGuiScreen(LegacyUiLaunchOption.clickGui);
    }

    private void updateStyle() {
        switch(styleValue.get().toLowerCase()) {
            case "liquidbounce":
                LegacyUiLaunchOption.clickGui.style = new LiquidBounceStyle();
                break;
            case "null":
                LegacyUiLaunchOption.clickGui.style = new NullStyle();
                break;
            case "slowly":
                LegacyUiLaunchOption.clickGui.style = new SlowlyStyle();
                break;
            case "black":
                LegacyUiLaunchOption.clickGui.style = new BlackStyle();
                break;
            case "white":
                LegacyUiLaunchOption.clickGui.style = new WhiteStyle();
                break;
        }
    }

    @EventTarget(ignoreCondition = true)
    public void onPacket(final PacketEvent event) {
        final Packet packet = event.getPacket();

        if (packet instanceof S2EPacketCloseWindow && mc.currentScreen instanceof ClickGui) {
            event.cancelEvent();
        }
    }
}

File: src/main/java/net/ccbluex/liquidbounce/ui/clickgui/ClickGUIModule.kt
/*
 * FDPClient Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge by LiquidBounce.
 * https://github.com/SkidderMC/FDPClient/
 */
package net.ccbluex.liquidbounce.ui.clickgui

import me.zywl.fdpclient.event.EventTarget
import me.zywl.fdpclient.event.PacketEvent
import net.ccbluex.liquidbounce.features.module.Module
import net.ccbluex.liquidbounce.features.module.ModuleCategory
import net.ccbluex.liquidbounce.features.module.ModuleInfo
import net.ccbluex.liquidbounce.ui.clickgui.style.styles.AstolfoStyle
import net.ccbluex.liquidbounce.ui.clickgui.style.styles.BlackStyle
import net.ccbluex.liquidbounce.ui.clickgui.style.styles.MixedStyle
import net.ccbluex.liquidbounce.ui.clickgui.style.styles.NullStyle
import net.ccbluex.liquidbounce.ui.clickgui.style.styles.classic.DropdownGUI
import net.ccbluex.liquidbounce.ui.clickgui.style.styles.fdpdropdown.FDPDropdownClickGUI
import net.ccbluex.liquidbounce.ui.gui.colortheme.ClientTheme
import net.ccbluex.liquidbounce.ui.gui.options.modernuiLaunchOption
import net.ccbluex.liquidbounce.value.BoolValue
import net.ccbluex.liquidbounce.value.FloatValue
import net.ccbluex.liquidbounce.value.IntegerValue
import net.ccbluex.liquidbounce.value.ListValue
import net.minecraft.network.play.server.S2EPacketCloseWindow
import org.lwjgl.input.Keyboard
import java.awt.Color
import java.util.*

@ModuleInfo(name = "ClickGUI", category = ModuleCategory.CLIENT, keyBind = Keyboard.KEY_RSHIFT, canEnable = false)
object ClickGUIModule : Module() {

    private val styleValue: ListValue = object : ListValue(
        "Style",
        arrayOf(
            "Classic",
            "Dropdrown",
//            "LB+",
            "Astolfo",
            "Mixed",
            "Null",
            "Black",
        ),
        "Dropdrown"
    ) {
        override fun onChanged(oldValue: String, newValue: String) {
            updateStyle()
        }
    }

    val scaleValue = FloatValue("Scale", 1f, 0.7f, 2f)
    val maxElementsValue = IntegerValue("MaxElements", 20, 1, 35)
    var clientColor = BoolValue("Client Color", true)
    val colorRedValue = IntegerValue("R", 0, 0, 255).displayable { !clientColor.get() } as IntegerValue
    val colorGreenValue = IntegerValue("G", 160, 0, 255).displayable { !clientColor.get() } as IntegerValue
    val colorBlueValue = IntegerValue("B", 255, 0, 255).displayable { !clientColor.get() } as IntegerValue
    val fastRenderValue = BoolValue("FastRender", false)
    val getClosePrevious = BoolValue("ClosePrevious", false)

    val backback = BoolValue("Background Accent", true)
    val scrollMode = ListValue("Scroll Mode", arrayOf("Screen Height", "Value"), "Value")
    val colormode = ListValue("Setting Accent", arrayOf("White", "Color"), "Color")
    val clickHeight = IntegerValue("Tab Height", 250, 100, 500)

    private var dropdown = DropdownGUI()
    private var dropdownClickGUIFDP = FDPDropdownClickGUI()

    override fun onEnable() {
        when {
            styleValue.get().equals("Classic", ignoreCase = true) -> mc.displayGuiScreen(dropdown)
            styleValue.get().equals("Dropdrown", ignoreCase = true) -> mc.displayGuiScreen(dropdownClickGUIFDP)
//            styleValue.get().equals("LB+", ignoreCase = true) -> mc.displayGuiScreen(NewUi.getInstance())
            else -> {
                updateStyle()
                mc.displayGuiScreen(modernuiLaunchOption.clickGui)
            }
        }
        this.state = false
    }
    private fun updateStyle() {
        when (styleValue.get().lowercase(Locale.getDefault())) {
            "null" -> modernuiLaunchOption.clickGui.style = NullStyle()
            "black" -> modernuiLaunchOption.clickGui.style = BlackStyle()
            "mixed" -> modernuiLaunchOption.clickGui.style = MixedStyle()
            "astolfo" -> modernuiLaunchOption.clickGui.style = AstolfoStyle()
        }
    }

    @JvmStatic
    fun generateColor(index : Int) : Color {
        return if (clientColor.get()) {
            ClientTheme.getColor(index)
        } else {
            Color(colorRedValue.get(), colorGreenValue.get(), colorBlueValue.get())
        }
    }

    @EventTarget(ignoreCondition = true)
    fun onPacket(event: PacketEvent) {
        val packet = event.packet
        if (packet is S2EPacketCloseWindow && mc.currentScreen is ClickGui) {
            event.cancelEvent()
        }
    }
}
